FireAxe Upstream (#1732)
firesim/firesim: docs/Advanced-Usage/FireAxe-Partitioning-onto-Multiple-FPGAs/Running-Fast-Mode-Simulations.rst

.. _FIREAXE-FAST-MODE:

Running Fast Mode Simulations
===================================

In this section, we provide a step-by-step tutorial on how to partition a
*RocketTile* out from the *SoC* and run fast-mode simulations using 2 FPGAs on EC2 F1.
This assumes that you have completed the :ref:`AWS F1 Getting Started Guide<AWS-F1-Getting-Started-Guide>` and that you are familiar with
the FireSim workflow.

> 1. Building Partitioned Sims : Setting up FireAxe Target configs

Reviewer: Nit: Everywhere you use `WORD : WORD`. To keep this consistent with the rest of the docs remove the 1st space `WORD: WORD`.

----------------------------------------------------------------

To build bitstreams for partitioned simulations, we need specify which modules we want to partition out. 

> We can do that in `sim/firesim-lib/src/main/scala/configs/FireAxeTargetConfigs.scala`.

Reviewer: ```suggestion
We can do that in :gh-file-ref:`sim/firesim-lib/src/main/scala/configs/FireAxeTargetConfigs.scala`.
```

.. literalinclude:: ../../../sim/firesim-lib/src/main/scala/configs/FireAxeTargetConfigs.scala
   :language: scala
   :start-after: DOC include start: F1 Rocket Partition
   :end-before: DOC include end: F1 Rocket Partition

> The `WithPartitionGlobalInfo` fragment takes in an argument of type `Seq[Seq[String]]`

Reviewer: Code formatting requires "``" instead of "`" everywhere. This needs fixing everywhere.

where each `Seq[String]` indicates the group of modules to be extracted out to the same partition.

.. code-block:: scala

    new WithPartitionGlobalInfo(Seq(
      Seq("A", "B"),
      Seq("C", "D"))

For instance, in the above example, modules "A", "B" will be grouped and extracted out to one FPGA,
"C", "D" will be grouped and extracted out to another FPGA, and the base SoC will be placed on the third FPGA.

As in the above example, we can use `WithPartitionBase` and `WithPartitionIndex(idx)` to specify the partitions.
`WithPartitionIndex(0)` for example, will choose `Seq("A", "B")` as the group of modules to partition out.

> How each partition group is mapped to the FPGA is specified in `user_topoloy.py`.

Reviewer: ```suggestion
How each partition group is mapped to the FPGA is specified in :gh-file-ref:`deploy/runtools/user_topology.py`.
```

2. Building Partitioned Sims : `config_build_recipes.yaml`
--------------------------------------------------------

Now, we can specify the `config_build_recipes.yaml`.
We need to set the `TARGET_CONFIG` to the FireChip configuration that we want to build, and the `PLATFORM_CONFIG` to the
partition configs that we defined in the above step.

.. literalinclude:: ../../../deploy/sample-backup-configs/config_build_recipes.yaml
   :language: yaml
   :start-after: DOC include start: F1 Rocket Partition Build Recipe
   :end-before: DOC include end: F1 Rocket Partition Build Recipe

You can now use the `firesim buildbitstream` command to build bitstreams of these partition configurations.

3. Running Partitioned Simulations : `user_topology.py`
--------------------------------------------------------

Once the bitstreams are built and you copied the hwdb entries into `config_hwdb.yaml`,

> we need to setup `config_runtime.yaml` and `user_topology.py` to run FireAxe simulations.

Reviewer: ```suggestion
we need to setup `config_runtime.yaml` and :gh-file-ref:`deploy/runtools/user_topology.py` to run FireAxe simulations.
```
